flextape/service/queue.go

```diff
@@ -0,0 +1,137 @@
+package service
+
+import (
+	"sort"
+)
+
+// invocationq is a simple queue of invocations.
+//
+// Beside implementing the classic Enqueue and Dequeue methods:
+// - It partially implements the sort.Interface to allow the easy
+//   implementation of invocation prioritization policies.
+// - It maintains per-entry state (QueueID) so it's easy to tell
+//   the position of each invocation by just looking at that value.
+type invocationq []*invocation
+
+func (iq *invocationq) Enqueue(x *invocation) Position {
+	position := Position(len(*iq))
+	offset := QueueID(1)
+	if len(*iq) > 0 {
+		offset = (*iq)[0].QueueID
+	}
+
+	x.QueueID = offset + QueueID(position)
+	(*iq) = append(*iq, x)
+	return position
+}
+
+func (iq *invocationq) Dequeue() *invocation {
+	if len(*iq) <= 0 {
+		return nil
+	}
+
+	retval := (*iq)[0]
+	(*iq) = (*iq)[1:len(*iq)]
+	retval.QueueID = 0
+	return retval
+}
+
+func (iq *invocationq) Len() int {
+	return len(*iq)
+}
+
+func (iq *invocationq) Swap(i, j int) {
+	(*iq)[i].QueueID, (*iq)[j].QueueID = (*iq)[j].QueueID, (*iq)[i].QueueID
+	(*iq)[i], (*iq)[j] = (*iq)[j], (*iq)[i]
+}
+
+// Filter is a function used to remove entries from the queue.
+//
+// It is invoked once per invocation in the queue, returns true if the entry
+// has to be filtered (removed), false if the entry has to be kept.
+type Filter func(pos Position, inv *invocation) bool
+
+// Walker is a function that does something for each entry in the queue.
+//
+// It is invoked once per invocation in the queue, returns true if the walk
+// should continue, false if it should stop.
+type Walker func(pos Position, inv *invocation) bool
+
+func (iq *invocationq) Filter(filter Filter) int {
+	newQueue := []*invocation{}
+	count := 0
+	for pos, inv := range *iq {
+		if filter(Position(pos+1), inv) {
+			count += 1
+			continue
+		}
+
+		inv.QueueID -= QueueID(count)
+		newQueue = append(newQueue, inv)
+	}
+	(*iq) = newQueue
+	return count
+}
+
+func (iq *invocationq) Walk(walker Walker) (*invocation, Position) {
```

Comment:
Public functions need docstrings (I got confused between this and `Walker` above)